tools/crm_rule.c

```diff
@@ -0,0 +1,296 @@
+
+    {"-spacer-",   no_argument,       NULL, '-', "\nData:"},
+    {"xml-text",   required_argument, NULL, 'X', "Use argument for XML (or stdin if '-')"},
+
+    {"-spacer-",   required_argument, NULL, '-', "\n\nThis tool is currently experimental.",
+     pcmk_option_paragraph},
+    {"-spacer-",   required_argument, NULL, '-', "The interface, behavior, and output may "
+                                                 "change with any version of pacemaker.", pcmk_option_paragraph},
+
+    {0, 0, 0, 0}
+};
+
+static int
+crm_rule_check(pe_working_set_t *data_set, const char *rule_id, crm_time_t *effective_date)
+{
+    xmlNode *cib_constraints = NULL;
+    xmlNode *match = NULL;
+    xmlXPathObjectPtr xpathObj = NULL;
+    pe_eval_date_result_t result;
+    char *xpath = NULL;
+    int rc = pcmk_ok;
+    int max = 0;
+
+    /* Rules are under the constraints node in the XML, so first find that. */
+    cib_constraints = get_object_root(XML_CIB_TAG_CONSTRAINTS, data_set->input);
+
+    /* Get all rules matching the given ID, which also have only a single date_expression
+     * child whose operation is not 'date_spec'.  This is fairly limited, but it's hard
+     * to check expressions more complicated than that.
+     */
+    xpath = crm_strdup_printf("//rule[@id='%s']/date_expression[@operation!='date_spec']", rule_id);
+    xpathObj = xpath_search(cib_constraints, xpath);
+
+    max = numXpathResults(xpathObj);
+    if (max == 0) {
+        CMD_ERR("No rule found with ID=%s containing a date_expression", rule_id);
+        rc = -ENXIO;
+        goto bail;
+    } else if (max > 1) {
+        CMD_ERR("More than one date_expression in %s is not supported", rule_id);
+        rc = -EOPNOTSUPP;
+        goto bail;
+    }
+
+    match = getXpathResult(xpathObj, 0);
+
+    /* We should have ensured both of these pass with the xpath query above, but
+     * double checking can't hurt.
+     */
+    CRM_ASSERT(match != NULL);
+    CRM_ASSERT(find_expression_type(match) == time_expr);
+
+    result = pe_eval_date_expression(match, effective_date);
+
+    if (result == pe_date_within_range) {
+        printf("Rule %s is still in effect\n", rule_id);
+        rc = 0;
+    } else if (result == pe_date_after_range) {
+        printf("Rule %s is expired\n", rule_id);
+        rc = 1;
+    } else if (result == pe_date_before_range) {
+        printf("Rule %s has not yet taken effect\n", rule_id);
+        rc = 2;
+    } else {
+        printf("Could not determine whether rule %s is expired\n", rule_id);
+        rc = 3;
+    }
+
+bail:
+    free(xpath);
+    freeXpathObject(xpathObj);
+    return rc;
+}
+
+int
+main(int argc, char **argv)
+{
+    cib_t *cib_conn = NULL;
+    pe_working_set_t *data_set = NULL;
+
+    int flag = 0;
+    int option_index = 0;
+
+    char *rule_id = NULL;
+    crm_time_t *rule_date = NULL;
+
+    xmlNode *input = NULL;
+    char *input_xml = NULL;
+    char *input_file = getenv("CIB_file");
```

Comment:
It's not necessary to look for CIB_file explicitly -- the CIB API handles it transparently. You can just remove all the file-specific handling and use the live CIB functions. If CIB_file is set in the environment, the CIB signon() and query() functions will open and read from the file rather than contact the CIB daemon.